key_management/src/wallet.rs

```diff
@@ -0,0 +1,344 @@
+    fn from(key_info: KeyInfo) -> Self {
+        let public_key =
+            wallet_helpers::to_public(*key_info.key_type(), key_info.private_key()).unwrap();
+        let address = wallet_helpers::new_address(*key_info.key_type(), &public_key).unwrap();
+        Key {
+            key_info,
+            public_key,
+            address,
+        }
+    }
+}
+
+/// This is a Wallet, it contains 2 HashMaps:
+/// - keys which is a HashMap of Keys resolved by their Address
+/// - keystore which is a HashMap of KeyInfos resolved by their Address
+#[derive(Clone, PartialEq, Debug, Eq)]
+pub struct Wallet<T> {
+    keys: HashMap<Address, Key>,
+    keystore: T,
+}
+
+impl Wallet<MemKeyStore> {
+    /// Return a wallet from a given amount of keys. This wallet will not use the
+    /// generic keystore trait, but rather specifically use a MemKeyStore
+    pub fn new_from_keys(key_vec: impl IntoIterator<Item = Key>) -> Self {
+        let mut keys: HashMap<Address, Key> = HashMap::new();
+        for item in key_vec.into_iter() {
+            keys.insert(item.address, item);
+        }
+        let key_store = MemKeyStore::new();
+        Wallet {
+            keys,
+            keystore: key_store,
+        }
+    }
+}
+
+impl<T> Wallet<T>
+where
+    T: KeyStore,
+{
+    /// Return a new Wallet with a given KeyStore
+    pub fn new(keystore: T) -> Self {
+        Wallet {
+            keys: HashMap::new(),
+            keystore,
+        }
+    }
+
+    /// Return the Key that is resolved by a given Address,
+    /// If this key does not exist in the keys hashmap, check if this key is in
+    /// the keystore, if it is, then add it to keys, otherwise return Error
+    pub fn find_key(&mut self, addr: &Address) -> Result<Key, Error> {
+        if let Some(k) = self.keys.get(&addr) {
+            return Ok(k.clone());
+        }
+        let key_string = format!("wallet-{}", addr.to_string());
+        let key_info = self.keystore.get(&key_string)?;
+        let new_key = Key::from(key_info.clone());
+        self.keys.insert(*addr, new_key.clone());
+        Ok(new_key)
+    }
+
+    /// Return the resultant Signature after signing a given message
+    pub fn sign(&mut self, addr: &Address, msg: &[u8]) -> Result<Signature, Error> {
+        // this will return an error if the key cannot be found in either the keys hashmap or it
+        // is not found in the keystore
+        let key = self.find_key(addr).map_err(|_| Error::KeyNotExists)?;
+        wallet_helpers::sign(*key.key_info.key_type(), key.key_info.private_key(), msg)
+    }
+
+    /// Return the KeyInfo for a given Address
+    pub fn export(&mut self, addr: &Address) -> Result<KeyInfo, Error> {
+        let k = self.find_key(addr)?;
+        Ok(k.key_info)
+    }
+
+    /// Add Key_Info to the Wallet, return the Address that resolves to this newly added KeyInfo
+    pub fn import(&mut self, key_info: &KeyInfo) -> Result<Address, Error> {
+        let k = Key::from(key_info.clone());
```

Comment:
```suggestion
    pub fn import(&mut self, key_info: KeyInfo) -> Result<Address, Error> {
        let k = Key::from(key_info);
```
This should be moved in to avoid always cloning if possible